README.md
# Test Upstream

https://hub.docker.com/r/ericof/test-upstream/


CHANGELOG.md
### 0.2.0

- Add sleep route (ericof)
- Add error route (ericof)

### 0.1.0

- Initial release (ericof)

tests/tests.py
# -*- coding: utf-8 -*-
"""Test upstream responses."""
import requests
import os
import unittest


class UpstreamTestCase(unittest.TestCase):
    """Testcase for upstream docker container."""

    def setUp(self):
        """Setup testcase."""
        self.baseurl = 'http://{host}:{port}'.format(
            host=os.environ.get('UPSTREAM_IP', '127.0.0.1'),
            port=os.environ.get('UPSTREAM_PORT', '8080')
        )

    def get_url(self, endpoint):
        """Return the url for a given endpoint."""
        return '{base_url}{endpoint}'.format(
            base_url=self.baseurl,
            endpoint=endpoint
        )

    def test_get(self):
        """Test GET verb."""
        endpoint = '/foo'
        resp = requests.get(
            self.get_url(endpoint)
        )
        self.assertEqual('application/json', resp.headers['Content-Type'])
        self.assertEqual(200, resp.status_code)
        self.assertEqual(endpoint, resp.json()['route'])
        self.assertEqual('GET', resp.json()['verb'])

        endpoint = '/foo/bar/foobar/barfoo'
        resp = requests.get(
            self.get_url(endpoint)
        )
        self.assertEqual(endpoint, resp.json()['route'])

    def test_get_query_string(self):
        """Test GET verb with query string."""
        endpoint = '/foo/bar/?foobar=moo&barfoo=ca'
        resp = requests.get(
            self.get_url(endpoint)
        )
        self.assertEqual('application/json', resp.headers['Content-Type'])
        self.assertEqual(200, resp.status_code)
        self.assertEqual('/foo/bar/', resp.json()['route'])
        self.assertIsInstance(resp.json()['query_data'], dict)
        self.assertEqual(resp.json()['query_data']['foobar'], 'moo')

    def test_post(self):
        """Test POST verb."""
        endpoint = '/foo'
        payload = {'some': 'data'}
        resp = requests.post(
            self.get_url(endpoint),
            data=payload
        )
        self.assertEqual('application/json', resp.headers['Content-Type'])
        self.assertEqual(200, resp.status_code)
        self.assertEqual(endpoint, resp.json()['route'])
        self.assertEqual('POST', resp.json()['verb'])
        self.assertIsInstance(resp.json()['form_data'], dict)

        endpoint = '/foo/bar/foobar/barfoo'
        resp = requests.post(
            self.get_url(endpoint),
            data=payload
        )
        self.assertEqual(endpoint, resp.json()['route'])

    def test_sleep(self):
        """Test sleep route with GET."""
        endpoint = '/sleep/1'
        resp = requests.get(
            self.get_url(endpoint)
        )
        self.assertEqual('application/json', resp.headers['Content-Type'])
        self.assertEqual(200, resp.status_code)
        self.assertEqual(endpoint, resp.json()['route'])
        self.assertEqual('GET', resp.json()['verb'])

    def test_error_route(self):
        """Test sleep route with GET."""
        endpoint = '/error_route/404'
        resp = requests.get(self.get_url(endpoint))
        self.assertEqual(404, resp.status_code)

        endpoint = '/error_route/500'
        resp = requests.get(self.get_url(endpoint))
        self.assertEqual(404, resp.status_code)

        endpoint = '/error_route/502'
        resp = requests.get(self.get_url(endpoint))
        self.assertEqual(404, resp.status_code)

        endpoint = '/error_route/503'
        resp = requests.get(self.get_url(endpoint))
        self.assertEqual(404, resp.status_code)

        endpoint = '/error_route/504'
        resp = requests.get(self.get_url(endpoint))
        self.assertEqual(404, resp.status_code)


tests/test_config.py
"""Test upstream responses."""
import pytest
import requests

from prettyconf import config


@pytest.fixture
def base_url():
    host = config('UPSTREAM_IP', default='127.0.0.1')
    port = config('UPSTREAM_PORT', default='8080')
    return f'http://{host}:{port}'


test_data = [
    '/foo',
    '/foo/bar/foobar/barfoo',
]

@pytest.mark.parametrize('endpoint', test_data)
def test_get(base_url, endpoint):
    """Test GET verb."""
    resp = requests.get(f'{base_url}{endpoint}')
    assert resp.headers['Content-Type'] == 'application/json'
    assert resp.status_code == 200
    assert resp.json()['route'] == endpoint
    assert resp.json()['verb'] == 'GET'


test_data = [
    ('/foo/bar/?foobar=moo&barfoo=ca', '/foo/bar/', 'foobar', 'moo'),
    ('/foo/bar/?foobar=moo&barfoo=ca', '/foo/bar/', 'barfoo', 'ca'),
]

@pytest.mark.parametrize('endpoint,route,key,value', test_data)
def test_get_query_string(base_url, endpoint, route, key, value):
    """Test GET verb with query string."""
    resp = requests.get(f'{base_url}{endpoint}')
    assert resp.headers['Content-Type'] == 'application/json'
    assert resp.status_code == 200
    assert resp.json()['route'] == route
    assert resp.json()['verb'] == 'GET'
    assert isinstance(resp.json()['query_data'], dict) is True
    assert resp.json()['query_data'][key] == value

test_data = [
    ('/foo/', {'key': 'value'}),
    ('/foo/bar', {'key': 'value'}),
]
@pytest.mark.parametrize('endpoint,payload', test_data)
def test_post(base_url, endpoint, payload):
    resp = requests.post(f'{base_url}{endpoint}', data=payload)
    assert resp.headers['Content-Type'] == 'application/json'
    assert resp.status_code == 200
    assert resp.json()['route'] == endpoint
    assert resp.json()['verb'] == 'POST'
    assert isinstance(resp.json()['form_data'], dict) is True

test_data = [
    '/sleep/1',
    '/sleep/2',
]
@pytest.mark.parametrize('endpoint', test_data)
def test_sleep(base_url, endpoint):
    """Test sleep route with GET."""
    resp = requests.get(f'{base_url}{endpoint}')
    assert resp.headers['Content-Type'] == 'application/json'
    assert resp.status_code == 200
    assert resp.json()['route'] == endpoint
    assert resp.json()['verb'] == 'GET'


test_data = [
    ('/error/404', 404),
    ('/error/500', 500),
    ('/error/502', 502),
    ('/error/503', 503),
    ('/error/504', 504),
]
@pytest.mark.parametrize('endpoint,status_code', test_data)
def test_error_route(base_url, endpoint, status_code):
    """Test sleep route with GET."""
    resp = requests.get(f'{base_url}{endpoint}')
    assert resp.status_code == status_code


echo.py
# -*- coding: utf-8 -*-
"""Simple echo chamber service."""
import bottle
import json

from prettyconf import config
from time import sleep


PORT = config('PORT', default=8080)


@bottle.route('/sleep/<how_long:int>', method='ANY')
def sleep_route(how_long):
    """Sleep for the time given as parameter
       then return a JSON response with the same url.
    """
    request = bottle.request
    method = request.method
    headers = request.headers
    headers_data = dict([(key, headers[key]) for key in headers])
    query_data = dict([(key, request.query[key]) for key in request.query])
    form_data = dict([(key, request.forms[key]) for key in request.forms])
    body = {
        'route': f'/sleep/{how_long}',
        'verb': method,
        'headers': headers_data,
        'form_data': form_data,
        'query_data': query_data,
    }

    return bottle.HTTPResponse(
        status=200,
        content_type='application/json',
        body=json.dumps(body)
    )


@bottle.route('/error/<status_code:int>', method='ANY')
def error_route(status_code):
    """Return the status code."""
    request = bottle.request
    method = request.method
    headers = request.headers
    headers_data = dict([(key, headers[key]) for key in headers])
    query_data = dict([(key, request.query[key]) for key in request.query])
    form_data = dict([(key, request.forms[key]) for key in request.forms])
    body = {
        'route': f'/error/{status_code}',
        'verb': method,
        'headers': headers_data,
        'form_data': form_data,
        'query_data': query_data,
    }

    return bottle.HTTPResponse(
        status=status_code,
        content_type='application/json',
        body=json.dumps(body)
    )


@bottle.route('<name:path>', method='ANY')
def catch_all(name):
    """Catch all requests, return a JSON response with the same url."""
    name = name if name.startswith('/') else '/{}'.format(name)
    request = bottle.request
    method = request.method
    headers = request.headers
    headers_data = dict([(key, headers[key]) for key in headers])
    query_data = dict([(key, request.query[key]) for key in request.query])
    form_data = dict([(key, request.forms[key]) for key in request.forms])
    body = {
        'route': name,
        'verb': method,
        'headers': headers_data,
        'form_data': form_data,
        'query_data': query_data,
    }

    return bottle.HTTPResponse(
        status=200,
        content_type='application/json',
        body=json.dumps(body)
    )

bottle.run(host='0.0.0.0', port=PORT)
